Add new virtual zarr page and remove dedicated kerchunk page (#189)
cloudnativegeo/cloud-optimized-geospatial-formats-guide: zarr/virtual-zarr.qmd

---
title: "Virtual Zarr"
subtitle: Zarr interface for data in archival formats
aliases:
  - /kerchunk/intro.html
---

# What is virtual Zarr?

> Virtual Zarr lets you interact with chunked archival data as if it were Zarr. You can get the benefits of Zarr without ever duplicating the data. 

Reviewer: ```suggestion
Virtual Zarr lets you interact with chunked archival data as if it were Zarr. You can get the benefits of Zarr without ever duplicating the data, including:

- partial reads, which minimize data transfer
- parallel reads, which speed up data loading
- aggregated metadata, which make it faster to inspect the contents of a dataset

```

Virtual Zarr stores capture metadata about where particular chunks of data are stored and allows tools like xarray to access the data from the chunks directly. These virtual Zarr stores can be written to disk using Kerchunk or Icechunk and distributed alongside data.

## How does it relate to regular Zarr?

> Zarr can be separated into two separate specifications. There is the Zarr interface which describes how libraries such as xarray interact with Zarr Arrays and Stores. That Zarr interface is implemented in libraries such as zarr-python, zarrs, and zarrita.js. Then there is the storage specification. When we think of Zarr stores we normally think of "native" Zarr Stores where the data is stored alongside the metadata. They are in the same hierarchy and sit on the same bucket on cloud storage.

Reviewer: ```suggestion
Zarr can be separated into two separate components. There is the Zarr interface (a.k.a. API) which describes how libraries such as xarray interact with Zarr Arrays and Groups. That Zarr interface is implemented in libraries such as zarr-python, zarrs, and zarrita.js. Then there is the storage component, meaning how the data persists outside a computers memory (on cloud storage, a hard drive, etc.). When we think of Zarr stores we normally think of "native" Zarr Stores where the data is stored alongside the metadata. They are in the same hierarchy and sit on the same bucket on cloud storage.
```

a few changes:
- two separate specifications -> two separate components, since the future spec committee has yet to resolve whether the zarr spec defines the at-rest condition.
- add synonym of API with interface
- change Zarr Stores -> Groups

Virtual Zarr Stores implement the Zarr interface, but instead of storing the chunk data in the hierarchy the interface sits on top of data stored in archival data formats.

::: {.callout-warning}

> A major limitation of virtual Zarr stores is that the chunking of data will always be constrained to the chunk structure of the underlying data format. With virtualization you can never change the chunks or the compression of the original data.

Reviewer: ```suggestion
A major limitation of virtual Zarr stores is that the chunking of data will always be constrained to the chunk structure of the underlying data format. A chunk refers to the bytes in the original data file that are compressed as one unit. Those chunks cannot be changed during virtualization since those compressed units live outside the Virtual Zarr store.
```

I think it's important to define a chunk, because if you think of it as the minimal unit of compression it cannot be changed. It you think of it as a read unit it technically can be subchunked for uncompressed data or aggregated.

Author: Definitely important to define a chunk! Thank you!

:::

## How does it work?

Virtual Zarr stores contain an index for every chunk of data in a particular dataset. This index gives information about each array and includes the path, byte offset, and length of each chunk. A client library like xarray lazily opens the index as a Zarr store without accessing the underlying chunk data at all. At compute time the Zarr store uses an HTTP range requests to fetch just the bytes of data that make up a particular chunk of data.

## How to a create a virtual Zarr store

To create a virtual Zarr store you first have to figure out the mapping between byte ranges and data chunks. To figure this out you have to scan every file that will be included in the virtual Zarr store. In some cloud-optimized formats like COG the necessary information is all in the file header, but in other formats it can be sprinkled throughout the files. So, depending on the structure of your archival data files, this process can take some time.

> There are two libraries for creating virtual Zarr stores: Kerchunk and VirtualiZarr.

Reviewer: I think it would be worth adding a section on wrappers that simplify these libraries. For example, I recently learned from @betolink that earthaccess has a wrapper that makes the ~20 lines of code for virtualizing NASA data only take a couple lines.

Author: Hmm. Yeah I didn't want to put too much on tooling in here, but the examples did get verbose... probably the interesting bits for illustrating the behavior is just to show the inputs and the output.

### Kerchunk

> [`kerchunk`](https://fsspec.github.io/kerchunk) is a Python library which generates a set of references to data or URLs under a key value store that matches the Zarr spec. 

Reviewer: ```suggestion
[`kerchunk`](https://fsspec.github.io/kerchunk) is a Python library which generates a set of references to data or URLs under a key value store that matches the Zarr spec. You can use kerchunk to serialize your virtual zarr store to kerchunk-json or kerchunk-parquet.
```

> Here's an example:

Reviewer: would it be worth making these uv scripts for reproducibility? e.g., we're talking about making the parser protocol for VirtualiZarr take a store rather than a registry, which would be backwards compatible with a deprecation cycle. Eventually, the example below may break.

Author: No strong opinion. My purpose with these examples are really to illustrate the concepts so I don't think it's crucial that they run. I could take out the paths to make it more obvious that this is just for showing what the libarary API looks like.

Reviewer: sounds good, I agree it's fine to just leave them as-is. I'll mark this as resolved

### Icechunk

Icechunk is specification for storing Zarr data that provides git-like data versioning and safe concurrent commits to the same store. Regular Icechunk supports storing data alongside metadata like native Zarr. But Icechunk can also be used to store virtual Zarr stores where the metadata points to existing chunks of data within archival formats.

::: {.callout-note}
Icechunk is an open source file format that is free to use - just like COG or Zarr. It is primarily developed by the company Earthmover. Earthmover has a commercial product called Arraylake that is build on the Icechunk format.
:::

Icechunk offers several advantages over Kerchunk. The most compelling is that Icechunk stores the last-modified time of the data chunks, so it knows when the underlying data chunks have changed relative to when the virtual store was written. This means that Icechunk will raise an error at read time if the data has changed. For a full comparison see the [VirtualiZarr docs](https://virtualizarr.readthedocs.io/en/stable/faq.html#which-format-should-i-save-my-virtual-references-as).

## Performance compared to native Zarr

> When streaming data from chunks, performance depends on how the chunks have been compressed, how much throughput the tooling supports, and the latency between compute environment and object storage. None of those are related to the data format. So at the chunk-level, if the chunks are the same size and the compression is the same then the performance should be the same. This is true for any chunked format - Zarr, virtual Zarr, COG, HDF5.

Reviewer: ```suggestion
When streaming data from chunks, performance depends on the size of the chunks, how the chunks have been compressed, how close together the chunks are in a data file, how much throughput and concurrent decompression the tooling supports, and the latency between compute environment and object storage. None of those are related to the data format. So at the chunk-level, if the chunks are the same size and the compression is the same then the performance should be the same. This is true for any chunked format - Zarr, virtual Zarr, COG, HDF5.
```

I'll think about this more, but I'm not 100% sure that I agree. I think the page size concept in some files like HDF5 can have meaningful file-format specific impact on performance by impactful coalescence potential

Author: Totally! If you are coalescing then there are certainly differences, but isn't that covered by the bit you added about "how close together the chunks are in the data file"? I mean in the extreme case of virtualizing a native unsharded Zarr then you can't coalesce at all right?

Reviewer: > If you are coalescing then there are certainly differences, but isn't that covered by the bit you added about "how close together the chunks are in the data file"

I think this makes the first sentence comprehensive.

I'd agree with the rest of the paragraph if "None of those are related to the data format" and the second to last sentence was changed to "if the chunks are the same size/shape and the compression is the same then the performance should be roughly the same", to account for the other real but less significant drivers of performance.

I suspect you might not agree with removing "None of those are related to the data format", and I would support merging even if that sentence remains. I just think there are concrete differences between file format specifications related to chunk sizes, positions, and compressions that genuinely matter for the performance drivers mentioned in the first sentence. e.g., some formats support different compressions, some formats have limits on how big a chunk size can be, some formats limit the number of chunks per file, and some formats limit adjacency between chunks in a file.

Author: Oh I agree with all that! I know I hammer the "chunks same" == "performance same" thing a lot 😄 I just think that's a really big source of confusion where people think things are going to just "be faster" if they rewrite COG to Zarr so my primary objective is to combat that myth. But I don't mind dialing down the rhetoric here.

Author: I made the changes you suggested in [d3bb6e5](https://github.com/cloudnativegeo/cloud-optimized-geospatial-formats-guide/pull/189/commits/d3bb6e58aea34e0a19f85711c8ce815afa0f0a3e)

---
title: "Virtual Zarr"
subtitle: Zarr interface for data in archival formats
author: Julia Signell
aliases:
  - /kerchunk/intro.html
---

# What is virtual Zarr?

Virtual Zarr stores lets you interact with chunked archival data as if it were Zarr. You can get the benefits of Zarr without ever duplicating the data, including:

- partial reads, which minimize data transfer
- parallel reads, which speed up data loading

> - aggregated metadata, which make it faster to inspect the contents of a dataset

Reviewer 2: ```suggestion
- aggregated metadata, which make it faster to inspect the contents of a dataset (also known as "lazy loading")
```

> Virtual Zarr stores capture metadata about where particular chunks of data are stored and lets tools like [xarray](https://docs.xarray.dev/) access the data from the chunks directly. These virtual Zarr stores can be written to disk using Kerchunk or Icechunk and distributed alongside data.

Reviewer 2: ```suggestion
Virtual Zarr stores capture metadata about where particular chunks of data are stored, enabling tools like [xarray](https://docs.xarray.dev/) to access relevant data from the chunks directly. These virtual Zarr stores can be written to disk using Kerchunk or Icechunk and distributed alongside data.
```

# How does it relate to regular Zarr?

> [Zarr](https://zarr-specs.readthedocs.io/en/latest/v3/core/index.html) can be separated into two separate components. There is the Zarr interface (a.k.a. API) which describes how libraries such as xarray interact with Zarr Arrays and Groups. That Zarr interface is implemented in libraries such as [zarr-python](https://zarr.readthedocs.io/en/stable/), [zarrs](https://zarrs.dev/), and [zarrita.js](https://zarrita.dev/). Then there is the storage component, meaning how the data persists outside a computers memory (on cloud storage, a hard drive, etc.). When we think of Zarr stores we normally think of "native" Zarr stores where the data is stored alongside the metadata. They are in the same hierarchy and sit on the same bucket on cloud storage.

Reviewer 2: ```suggestion
[Zarr](https://zarr-specs.readthedocs.io/en/latest/v3/core/index.html) can be separated into two separate components. There is the Zarr specification (or interface) which describes how libraries such as xarray interact with Zarr Arrays and Groups. That Zarr interface is implemented in libraries such as [zarr-python](https://zarr.readthedocs.io/en/stable/), [zarrs](https://zarrs.dev/), and [zarrita.js](https://zarrita.dev/). Then there is the storage component, meaning how the data persists outside a computers memory (on cloud storage, a hard drive, etc.). For most of Zarr's history, Zarr stores referred to stores which adherred to both the specification and the on-disk storage format. We call these "native" Zarr stores: data is stored in the Zarr on-disk format alongside Zarr metadata. Native Zarr stores usually store data and metadata in the same bucket on cloud storage. Other storage protocols are supported but what is important is the Zarr data is located alongside the metadata, so addresses can be relative.
```

Author: I really like this careful explanation of native Zarr stores! Thanks!

> Virtual Zarr stores implement the Zarr interface, but instead of storing the chunk data in the hierarchy the interface sits on top of data stored in archival data formats.

Reviewer 2: ```suggestion
Virtual Zarr stores implement the Zarr interface, but instead of storing the chunk data in the hierarchy, the virtual store provides an "indirection layer", pointing to chunks of data stored in archival data formats.
```

Reviewer 2: This is Max's term

There are two libraries for creating virtual Zarr stores: [Kerchunk](https://fsspec.github.io/kerchunk/) and [VirtualiZarr](https://virtualizarr.readthedocs.io/en/stable/).

## Kerchunk

[Kerchunk](https://fsspec.github.io/kerchunk) is a Python library which generates a set of references to data or URLs under a key value store that matches the Zarr spec. You can use kerchunk to serialize your virtual zarr store to kerchunk-json or kerchunk-parquet.

Here's an example:

```python
import fsspec
import json
from kerchunk.hdf import SingleHdf5ToZarr

local_file = 'some_data.nc'
out_file = 'some_references.json'

# Instantiate the local file system with fsspec to save kerchunk's reference data as json.
fs = fsspec.filesystem('')
in_file = fs.open(local_file)

# The inline threshold adjusts the size below which binary blocks are included directly in the output.
# A higher inline threshold can result in a larger json file but faster loading time overally, since fewer requests are made.
h5chunks = SingleHdf5ToZarr(in_file, local_file, inline_threshold=300)
with fs.open(out_file, 'wb') as f:
    f.write(json.dumps(h5chunks.translate()).encode())
```

> ## VirtualiZarr

Reviewer 2: I think we need to change the headings here, having Kerchunk at the same level and right before VirtualiZarr may think that kerchunk and Virtualizarr are swappable. It may help to use this breakdown Max worked out in slack a while ago:

* **Specifications:** Zarr, Icechunk, and Kerchunk all have specifications.
* **Data format:** Zarr defines how data is contained in multiple files (for local usage) or objects (for cloud usage). 
* **Data interface:** The Zarr data interface describes how libraries can interact with data stored on disk.
* **Indirection layers:** Icechunk and kerchunk provide specifications for how to store data so that it can be treated as a Zarr store but for data stored in other locations (e.g., a NetCDF file)
* **Libraries:** [Many libraries](https://zarr.dev/implementations/) can read and write Zarr data, these are called Zarr implementations. The [icechunk](https://github.com/earth-mover/icechunk) library can read and write Icechunk's representation of Zarr data. Kerchunk can write Kerchunk's representation of Zarr data. VirtualiZarr is a library which helps parse metadata from archival formats and writes to Icechunk or Kerchunk.

Reviewer 2: Oh I see, you have identified that both Kerchunk and VirtualiZarr are _libraries_ which can be used to generate chunk manifest metadata in-memory. This organization makes sense now, but I think the list above or some variation of it, to identify and name the components of this system, may still be helpful?

Author: I am hesitant about having too many different framings on this one high-level page. Maybe this is an opportunity for a diagram? Max had suggested this one: https://link.excalidraw.com/readonly/nD6RMWAOx5c13WOV7E0X which I could see layering with additional labels to make the relationship clearer. 

Another option could be to set up the section better in the first paragraph of "How to a create a virtual Zarr store"

Reviewer 2: Ya I think the current layout makes sense now that I'm looking at it holistically. The diagram is great and we could add, if you like, the annotation "VirtualiZarr parsers", "VirtualiZarr writers" to the first and third arrows. But also find to leave it out or leave as-is.

Author: I kind of ended up deciding that the diagram doesn't align super well with the framing in this doc. But I am happy to be overridden.

Here's an example of just the parsing step:

```python
import icechunk
import obstore

from obspec_utils.registry import ObjectStoreRegistry

from virtualizarr import open_virtual_dataset
from virtualizarr.parsers import HDFParser

bucket = "s3://nex-gddp-cmip6"
path = "NEX-GDDP-CMIP6/ACCESS-CM2/ssp126/r1i1p1f1/tasmax/tasmax_day_ACCESS-CM2_ssp126_r1i1p1f1_gn_2015_v2.0.nc"
store = obstore.store.from_url(bucket, region="us-west-2", skip_signature=True)

registry = ObjectStoreRegistry({bucket: store})

parser = HDFParser()
vds = open_virtual_dataset(
  url=f"{bucket}/{path}",
  parser=parser,
  registry=registry,
  loadable_variables=[],
)
```

# Storage formats

> Creating a virtual Zarr store takes some time, but the index that you generate is valid as long as the underlying data stays the same. In order to make use of the virtual Zarr store and potentially share it with others, you need to serialize it to disk. If you are using Kerchunk to create your virtual Zarr stores then you will always serialize to Kerchunk format. If you are using VirtualiZarr then you can serialize to Kerchunk format or to Icechunk.

Reviewer 2: ```suggestion
Creating a virtual Zarr store takes time, but the index that you generate is valid as long as the underlying data stays the same. In order to make use of the virtual Zarr store and potentially share it with others, you need to serialize it to disk. If you are using Kerchunk to create your virtual Zarr stores then you will always serialize to Kerchunk format. If you are using VirtualiZarr then you can serialize to Kerchunk format or to Icechunk.
```

## Kerchunk

[Kerchunk](https://fsspec.github.io/kerchunk/) has two options for storage: kerchunk-json and kerchunk-parquet. A simple reference file in kerchunk-json for a NetCDF file might look like:

```json
{
  ".zgroup": "{\n    \"zarr_format\": 2\n}",
  ".zattrs": "{\n    \"Conventions\": \"UGRID-0.9.0\n\"}",
  "x/.zattrs": "{\n    \"_ARRAY_DIMENSIONS\": [\n        \"node\"\n ...",
  "x/.zarray": "{\n    \"chunks\": [\n        9228245\n    ],\n    \"compressor\": null,\n    \"dtype\": \"<f8\",\n  ...",
  "x/0": ["s3://bucket/path/file.nc", 294094376, 73825960]
}
```
The `["s3://bucket/path/file.nc", 294094376, 73825960]` is the key part, which says that to load the first chunk in the `x` dimension, the Zarr reader needs to fetch a byte range starting at `294094376` with a length of `73825960` bytes. 

For very large virtual Zarr stores, kerchunk-json files can get too big to be practical. That's where kerchunk-parquet comes in. It uses the same concepts as kerchunk-json, but stores the information in parquet which has the advantages of better compressibility and file partitioning.

## Icechunk

> [Icechunk](https://icechunk.io/en/stable/) is specification for storing Zarr data that provides git-like data versioning and safe concurrent commits to the same store. Regular Icechunk supports storing data alongside metadata like native Zarr. But Icechunk can also be used to store virtual Zarr stores where the metadata points to existing chunks of data within archival formats.

Reviewer 2: ```suggestion
[Icechunk](https://icechunk.io/en/stable/) is a specification for storing Zarr data that provides git-like data versioning and safe concurrent commits to the same store. Regular Icechunk supports storing data alongside metadata like native Zarr. But Icechunk can also be used to store virtual Zarr stores where the metadata points to existing chunks of data within archival formats.
```

::: {.callout-note}

> Icechunk is an open source file format that is free to use - just like COG or Zarr. It is primarily developed by the company Earthmover. Earthmover has a commercial product called Arraylake that is build on the Icechunk format.

Reviewer 2: ```suggestion
Icechunk is an open source file format that is free to use - just like COG or Zarr. It is primarily developed by the company Earthmover. Earthmover has a commercial product called Arraylake that is built on the Icechunk format.
```

:::

Icechunk offers several advantages over Kerchunk. The most compelling is that Icechunk stores the last-modified time of the data chunks, so it knows when the underlying data chunks have changed relative to when the virtual store was written. This means that Icechunk will raise an error at read time if the data has changed. For a full comparison see the [VirtualiZarr docs](https://virtualizarr.readthedocs.io/en/stable/faq.html#which-format-should-i-save-my-virtual-references-as).

# Performance compared to native Zarr

When streaming data from chunks, performance depends on the size of the chunks, how the chunks have been compressed, how close together the chunks are in a data file, how much throughput and concurrent decompression the tooling supports, and the latency between compute environment and object storage. So at the chunk-level, if the chunks are the same size and the compression is the same then the performance should be roughly the same. This is true for any chunked format - Zarr, virtual Zarr, COG, HDF5.

> The benefit that you get in virtual Zarr is that you always have the metadata in one defined location. This means that you can read the entire metadata for the whole dataset in one request. As a comparison, in native Zarr (without consolidated metadata) you need one request per array. So when lazily opening a dataset virtual Zarr stores only do one request whereas native Zarr stores do as many as they have arrays. Doing many requests is not necessarily slower, but making it fast requires more sophisticated tooling.

Reviewer 2: will this always be true? I feel like the real power is already having mapped logical data coordinates to byte ranges. I got the sense that very large manifests could and should be broken up. You can load metadata in parallel so this is less of an issue than just being able to load all the metadata about data coordinate --> byte ranges without having to read or stream entire files.

Author: Yeah maybe I need to qualify this a bit, but conceptually even if you end up splitting the manifest it should still be an improvement to be able to fetch metadata about more than one array at a time.

> I feel like the real power is already having mapped logical data coordinates to byte ranges.

I thought the mapping was between chunk index and byte range, not data coordinate and byte range. But maybe I'm wrong!

Reviewer 2: no i think you are right, I was conflating coordinates and indices 👍🏽

# Reading from virtual Zarr stores

In principle, reading from virtual Zarr stores is just like reading from "native" Zarr stores. So once you have a virtual Zarr store in memory (in VirtualiZarr this is called a `ManifestStore`) you can use that store just as you would any other Zarr store:

```python
xr.open_zarr(manifest_store)
```

If you have a virtual Zarr store persisted to disk as Icechunk or Kerchunk then you use those libraries to read the virtual Zarr store. So for kerchunk:

```python
xr.open_dataset("s3://path/to/kerchunk.json", engine="kerchunk")

> For icechunk, the user explicitly specifies the authentication mechanism required for both the archival data loaction and icechunk location. This makes the syntax more verbose:

Reviewer 2: ```suggestion
For icechunk, the user explicitly specifies the authentication mechanism required for both the archival data location and icechunk location. This makes the syntax more verbose:
```

Author: thank you for catching these! I need to turn on the spellchecker plugin clearly.
